File: Tempo_odnc/QuickSort.java
import java.io.*;
import java.util.ArrayList;
import java.util.List;

public class QuickSort {

    public static void main(String[] args) {
        String arquivoEntrada = "arq(1).txt";
        String arquivoSaida = "arq-saida.txt";

        List<Integer> numeros = lerNumeros(arquivoEntrada);
        if (numeros == null) {
            System.out.println("Erro ao ler o arquivo de entrada.");
            return;
        }

        long tempoInicio = System.currentTimeMillis();
        List<Integer> numerosOrdenados = quickSort(numeros);
        long tempoFim = System.currentTimeMillis();

        escreverNumeros(arquivoSaida, numerosOrdenados);

        long tempoGasto = tempoFim - tempoInicio;  // em ms
        long memoriaUsada = medirMemoria();  // em KB

        System.out.printf("Tempo gasto: %d ms%n", tempoGasto);
        System.out.printf("Memória usada: %d KB%n", memoriaUsada);
    }

    public static List<Integer> quickSort(List<Integer> lista) {
        if (lista.size() <= 1) {
            return lista;
        }
        int pivo = lista.get(lista.size() / 2);
        List<Integer> esquerda = new ArrayList<>();
        List<Integer> meio = new ArrayList<>();
        List<Integer> direita = new ArrayList<>();
        
        for (int num : lista) {
            if (num < pivo) {
                esquerda.add(num);
            } else if (num == pivo) {
                meio.add(num);
            } else {
                direita.add(num);
            }
        }

        List<Integer> resultado = new ArrayList<>();
        resultado.addAll(quickSort(esquerda));
        resultado.addAll(meio);
        resultado.addAll(quickSort(direita));

        return resultado;
    }

    public static List<Integer> lerNumeros(String nomeArquivo) {
        List<Integer> numeros = new ArrayList<>();
        try (BufferedReader br = new BufferedReader(new FileReader(nomeArquivo))) {
            String linha;
            while ((linha = br.readLine()) != null) {
                numeros.add(Integer.parseInt(linha.trim()));
            }
        } catch (IOException e) {
            e.printStackTrace();
            return null;
        }
        return numeros;
    }

    public static void escreverNumeros(String nomeArquivo, List<Integer> numeros) {
        try (BufferedWriter bw = new BufferedWriter(new FileWriter(nomeArquivo))) {
            for (int numero : numeros) {
                bw.write(String.valueOf(numero));
                bw.newLine();
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static long medirMemoria() {
        Runtime runtime = Runtime.getRuntime();
        runtime.gc();  // Solicitar a coleta de lixo
        long memoriaUsada = (runtime.totalMemory() - runtime.freeMemory()) / 1024;  // em KB
        return memoriaUsada;
    }
}

File: Tempo_odnc/QuickSort.py
import time
import psutil
import os

def quick_sort(lista):
    if len(lista) <= 1:
        return lista
    pivo = lista[len(lista) // 2]
    esquerda = [x for x in lista if x < pivo]
    meio = [x for x in lista if x == pivo]
    direita = [x for x in lista if x > pivo]
    return quick_sort(esquerda) + meio + quick_sort(direita)

def ler_numeros(nome_arquivo):
    if not os.path.isfile(nome_arquivo):
        raise FileNotFoundError(f"O arquivo {nome_arquivo} não existe.")
    with open(nome_arquivo, 'r') as arquivo:
        return [int(linha.strip()) for linha in arquivo]

def escrever_numeros(nome_arquivo, numeros):
    with open(nome_arquivo, 'w') as arquivo:
        for numero in numeros:
            arquivo.write(f"{numero}\n")

def medir_memoria():
    processo = psutil.Process(os.getpid())
    return processo.memory_info().rss / 1024  

if __name__ == "__main__":
    arquivo_entrada = 'arq(1).txt'  
    arquivo_saida = 'arq-saida.txt'
    
    try:
        numeros = ler_numeros(arquivo_entrada)
    except FileNotFoundError as e:
        print(e)
        exit(1)
    
    tempo_inicio = time.time()
    numeros_ordenados = quick_sort(numeros)
    tempo_fim = time.time()
    
    escrever_numeros(arquivo_saida, numeros_ordenados)
    
    tempo_gasto = (tempo_fim - tempo_inicio) * 1000  
    memoria_usada = medir_memoria()  
    
    print(f"Tempo gasto: {tempo_gasto:.2f} ms")
    print(f"Memória usada: {memoria_usada:.2f} KB")
